### ISSUE TITLE
 rsyslog-server stops writing messages to file after 30 sec

### ISSUE THREAD

**Issue by qw3r3wq** [rsyslog/rsyslog#2574]
### Expected behavior
receive and write all logs ASAP when messages reach server.

### Actual behavior
writes logs with delay even when separate log partition has NO activity at the moment.
example logs:
[root@all-logs ~]# tail -f -n 1000 /var/log/remote/compute-0/2018/03/21/mail.log | grep TESTAS
Mar 21 17:11:20 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 0
Mar 21 17:11:23 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 1
Mar 21 17:11:26 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 2
Mar 21 17:11:29 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 3
Mar 21 17:11:32 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 4
Mar 21 17:11:35 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 5
Mar 21 17:11:38 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 6
Mar 21 17:11:41 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 7
Mar 21 17:11:44 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 8
Mar 21 17:11:47 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 9
Mar 21 17:11:50 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 10
Mar 21 17:11:53 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 11
Mar 21 17:11:59 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 13
Mar 21 17:11:56 10.94.0.125[10.94.0.125] TESTAS: TESTAS mail.info  TEST Message to be written 12

Also please note, that Message 12 went to log later than 13...

### Steps to reproduce the behavior
I have only 12 hosts sending all logs to this VM.
[root@all-logs ~]# lscpu
Architecture:          x86_64
CPU(s):                4
[root@all-logs ~]# free -m
              total        used        free      shared  buff/cache   available
Mem:           7976         275        6813           8         886        7429
Swap:             0           0           0
[root@all-logs ~]# 
mount:
/dev/mapper/CLOUD-LOGS on /var/log/remote type btrfs (rw,nosuid,nodev,noexec,noatime,nodiratime,sync,seclabel,compress=lzo,space_cache,autodefrag,subvolid=5,subvol=/)

reprodusable with kernel 4.14 (with reiser4 and 3.0.10 with XFS)

### Environment
- rsyslog version: rsyslogd 8.33.1
- platform: CentOS 7.4
- for configuration questions/issues, include rsyslog.conf:

---
I know, that there should be 8.34 but it isn't in repo:

[conf.txt](https://github.com/rsyslog/rsyslog/files/1833937/conf.txt)
[Re rsyslog rsyslog-server stops receiving messages after 30 sec.txt](https://github.com/rsyslog/rsyslog/files/1834068/Re.rsyslog.rsyslog-server.stops.receiving.messages.after.30.sec.txt)
[yum.list.txt](https://github.com/rsyslog/rsyslog/files/1834069/yum.list.txt)

[rsyslog.zip.001.zip](https://github.com/rsyslog/rsyslog/files/1834095/rsyslog.zip.001.zip)
[rsyslog.zip.002.zip](https://github.com/rsyslog/rsyslog/files/1834096/rsyslog.zip.002.zip)
[rsyslog.zip.003.zip](https://github.com/rsyslog/rsyslog/files/1834097/rsyslog.zip.003.zip)

rsyslog.zip.001.zip rsyslog.zip.002.zip rsyslog.zip.003.zip < remove last .zip and extract to see full output of:
rsyslogd -dn -f /etc/rsyslog.conf &>/tmp/rsyslog.txt

**Comment by davidelang** [comment]
please configure impstats and have it write to a file directly, when you get in 
the mode of no logs being written, let's see what is shown in the pstats output. 
This could be that an output is blocked or something like that.

**Comment by qw3r3wq** [comment]
I have disabled sending to elastic, and it  looks like it is working now.
So now I need fixing elastic :) but I want to wait for a 24 hours... with working config, and I will configure to send messages tro elastic back.

**Comment by davidelang** [comment]
This implies that there is some problem with delivering messages to ES, and as a 
result a queue fills up and the messages are blocked.

David Lang

**Comment by qw3r3wq** [comment]
I am exited, that messages were going, I enabled back elastic, found some FW issues, fixed that, still not working, so used LANG=C audit2why </var/log/audit/auditd..log and it game me:
setsebool -P nis_enabled 1
Enabled it, still receiving logs to files!

How to configure impstats ? I just want to be sure.

**Comment by davidelang** [comment]
start with
https://www.rsyslog.com/doc/v8-stable/configuration/modules/impstats.html

Ask when you run into questions.

**Comment by qw3r3wq** [comment]
Awesome! will update tomorrow, now finishing with elastic and adding MORE logs to server to make it  work harder :) hope won't get any issue wit no logs.
It's a pity it do not write anything to log before stopping message write...

**Comment by qw3r3wq** [comment]
Hi, just writing to update.
I had issue to push rsyslog to elastic it was sendimg wromg type of message. So i have done many workarounds, yesterday moving it back to standard way, all worked good, and still working.

I would rephrase 'bug' of this issue, that I haven't seen any logs in messages, that some buffer is filling up, which causes dramatical impact on mesage dump to disk...

Sorry for typo, writing from tablet.

**Comment by davidelang** [comment]
On Thu, 22 Mar 2018, qw3r3wq wrote:

> Hi, just writing to update.
> I had issue to push rsyslog to elastic it was sendimg wromg type of message. So i have done many workarounds, yesterday moving it back to standard way, all worked good, and still working.
>
> I would rephrase 'bug' of this issue, that I haven't seen any logs in messages, that some buffer is filling up, which causes dramatical impact on mesage dump to disk...

Remember that rsyslog only knows your intent by what you configure in the config 
file. It can't know that if ES is down you want it to continue writing to local 
files unless you configure it to do so, and when you configure it to do so, you 
are configuring a queue size of how many messages you can queue for ES. When 
that queue fills up, it blocks.

The impstats output, which can trivially be configured to write to a separate 
file that is not affected by the main queue (or can be put in a different queue 
inside rsyslog) is where the status of the queues shows up, including what 
queues are blocked.

Rsyslog can't write a message saying that there is a problem with the output as 
a general thing, because when it's following your rules as to what to do with 
log messages, the log message saying that there is a problem gets stuck in the 
queues behind other messages.

Now, you can setup your config so that messages generated internally by rsyslog 
can bypass all the other queues and will always get written to disk, but that's 
not actually a common thing for people to do :-)

David Lang
